src/Microsoft.IdentityModel.Tokens/Validators.cs

```diff
@@ -129,40 +146,58 @@ public static void ValidateAudience(IEnumerable<string> audiences, SecurityToken
                     {
-                        if (string.Equals(validAudience, tokenAudience, StringComparison.Ordinal))
-                        {
-                            LogHelper.LogInformation(LogMessages.IDX10234, tokenAudience);
-                            return;
-                        }
-                    }
-                    else if (validationParameters.IgnoreTrailingSlashWhenValidatingAudience)
-                    {
-                        var length = (validAudience.Length == tokenAudience.Length + 1 && validAudience.EndsWith("/", StringComparison.InvariantCulture))
-                                        ? validAudience.Length - 1
-                                        : (tokenAudience.Length == validAudience.Length + 1 && tokenAudience.EndsWith("/", StringComparison.InvariantCulture))
-                                            ? tokenAudience.Length - 1
-                                            : -1;
-
-                        // the length of the audiences is different by more than 1 and neither ends in a "/"
-                        if (length == -1)
-                            continue;
-
-                        if (string.CompareOrdinal(validAudience, 0, tokenAudience, 0, length) == 0)
-                        {
-                            LogHelper.LogInformation(LogMessages.IDX10234, tokenAudience);
-                            return;
-                        }
+                        LogHelper.LogInformation(LogMessages.IDX10234, tokenAudience);
+                        return true;
                     }
                 }
             }
 
-            throw LogHelper.LogExceptionMessage(
-                new SecurityTokenInvalidAudienceException(LogHelper.FormatInvariant(LogMessages.IDX10214, Utility.SerializeAsSingleCommaDelimitedString(audiences), (validationParameters.ValidAudience ?? "null"), Utility.SerializeAsSingleCommaDelimitedString(validationParameters.ValidAudiences)))
-                    { InvalidAudience = Utility.SerializeAsSingleCommaDelimitedString(audiences) });
+            return false;
+        }
+
+        private static bool AudiencesMatch(TokenValidationParameters validationParameters, string tokenAudience, string validAudience)
+        {
+            if (validAudience.Length == tokenAudience.Length)
+            {
+                if (string.Equals(validAudience, tokenAudience, StringComparison.Ordinal))
+                {
+                    return true;
+                }
+            }
+            else if (validationParameters.IgnoreTrailingSlashWhenValidatingAudience && AudiencesMatchIgnoringTrailingSlash(tokenAudience, validAudience))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool AudiencesMatchIgnoringTrailingSlash(string tokenAudience, string validAudience)
+        {
+            int length;
```

Comment:
length can be assigned -1 here itself ?

Reply:
Good call, that is easier to read. Originally just taking a naïve conversion from the nested ternary